controller/uninstall_controller.go

```diff
@@ -0,0 +1,433 @@
+		replicaInformerSynced:     replicaInformer.Informer().HasSynced,
+		engineImageInformerSynced: engineImageInformer.Informer().HasSynced,
+		nodeInformerSynced:        nodeInformer.Informer().HasSynced,
+		daemonSetInformerSynced:   daemonSetInformer.Informer().HasSynced,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(500*time.Millisecond, 5*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		), "longhorn-uninstall"),
+	}
+
+	volumeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineInformer.Informer().AddEventHandler(c.controlleeHandler())
+	replicaInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineImageInformer.Informer().AddEventHandler(c.controlleeHandler())
+	nodeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	daemonSetInformer.Informer().AddEventHandler(c.controlleeHandler())
+	return c
+}
+
+func (c *UninstallController) controlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueControlleeChange() },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueControlleeChange() },
+		DeleteFunc: func(obj interface{}) { c.enqueueControlleeChange() },
+	}
+}
+
+func (c *UninstallController) enqueueControlleeChange() {
+	c.queue.AddRateLimited("uninstall")
+}
+
+func (c *UninstallController) Run() error {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	if !controller.WaitForCacheSync("longhorn uninstall", c.stopCh,
+		c.volumeInformerSynced, c.engineImageInformerSynced,
+		c.nodeInformerSynced, c.daemonSetInformerSynced) {
+
+		return fmt.Errorf("Failed to sync informers")
+	}
+
+	if err := c.checkPreconditions(); err != nil {
+		c.stop()
+		return err
+	}
+
+	startTime := time.Now()
+	logrus.Infof("Uninstalling...")
+	defer func() {
+		fields := logrus.Fields{"runtime": time.Now().Sub(startTime)}
+		logrus.WithFields(fields).Infof("Uninstallation completed.")
+	}()
+	go wait.Until(c.worker, time.Second, c.stopCh)
+
+	<-c.stopCh
+	return nil
+}
+
+func (c *UninstallController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *UninstallController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.uninstall()
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *UninstallController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logrus.Warn(errors.Wrapf(err, "worker error"))
+	c.queue.AddRateLimited(key)
+}
+
+func (c *UninstallController) uninstall() (err error) {
+	defer func() {
+		if c.status != "" && c.lastStatus != c.status {
+			logrus.Infof("worker done: %s", c.status)
+			c.lastStatus = c.status
+		}
+		c.status = ""
+	}()
+
+	if err = c.managerReady(); err == nil {
+		if err = c.deleteCRDs(); err != nil || c.status != "" {
```

Comment:
OK, I still don't understand why we need the status here.

You said in another comment:
>Usually, if there's some error (ex: resource changed conflicts), we want to requeue. If we always return an error (until controller reaches completion), we always requeue, causing infinite queue loop.

Yeah, if there is a conflict, we will requeue. But I cannot see why it will cause infinite queue loop? Because in the end, there won't be a conflict, since you're always getting the latest object, not the object when you was requeuing.

Reply:
If we requeue immediately after every worker execution, we would be in an infinite loop. The first worker execution would run, requeue, run, requeue, etc. Only in error scenarios should we requeue a worker's execution, otherwise we should wait for a CRD change event to trigger it.

We used a bool for flow control before. That muddied up the method signature, so we tried using error to manipulate flow control. That introduced the queue infinite loop (since every run was a failure) so a controller-wide status field was added to handle flow control scenario where worker succeeded and should terminate, but controller is _not_ done and should wait for subsequent change events.

Reply:
OK, now I understand it. I think it's better to revert back to the boolean... Probably with a comment.

The `c.status` or similar is not threadsafe (though we don't have more than one thread now).

The variable can be named e.g. `toWaitForUpdate`.